api/envoy/type/v3/ratelimit_strategy.proto

```diff
@@ -0,0 +1,85 @@
+syntax = "proto3";
+
+package envoy.type.v3;
+
+import "envoy/type/v3/ratelimit_unit.proto";
+import "envoy/type/v3/token_bucket.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.type.v3";
+option java_outer_classname = "RatelimitStrategyProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/type/v3;typev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#not-implemented-hide:]
+// [#protodoc-title: Rate Limit Strategies]
+
+message RateLimitStrategy {
+  // Choose between allow all and deny all.
+  enum BlanketRule {
+    ALLOW_ALL = 0;
+    DENY_ALL = 1;
+  }
+
+  // Limit the number of requests per second.
+  message RequestsPerTimeUnit {
+    // The number of requests per unit of time to allow. Requests over the limit
+    // are denied.
+    // If set to 0, deny all (equivalent to BlanketRule.DENY_ALL).
+    uint64 requests_per_time_unit = 1;
+
+    // The unit of time. Ignored when requests_per_time_unit is 0 (deny all).
+    RateLimitUnit time_unit = 2;
+  }
+
+  oneof strategy {
+    option (validate.required) = true;
+
+    // Allow or Deny the requests.
+    // If unset, allow all.
+    BlanketRule blanket_rule = 1;
+
+    // A number of requests per time unit, f.e. requests per second.
+    RequestsPerTimeUnit requests_per_time_unit = 2;
+
+    // Limit the requests by consuming tokens from the Token Bucket.
+    // Allow the same number of requests as the number of tokens available in
+    // the token bucket.
+    TokenBucket token_bucket = 3;
```

Comment:
`TokenBucket` and `RequestsPerTimeUnit` seems to be equivalent. Why do we need both?

Reply:
I'm was thinking of `RequestsPerTimeUnit` option as nonnopionated instruction to the client, something like "feel free to choose your own rate limiting algorithm". Practically this will identical to `TokenBucket` in Envoy and gRPC implementations, just less opinionated way to express it. This could be useful if we choose a different flavor of Token Bucket to be the default algorithm, f.e. Leaky Bucket, Dynamic Token Bucket. Do you think this is useful, or unnecessary superfluous?

Reply:
@yanavlasov friendly ping here.

Reply:
Sounds good, we can keep both

Reply:
I think you're right in the sense this needs explanation though. I'll capture this note in the doc.

Reply:
Added more explanations in https://github.com/envoyproxy/envoy/pull/19793/commits/075f7a5e9f526022eb55891414b1bad7b567a08d.